### DISCUSSION TITLE
How to use xmeshfem2D with external model ?

### DISCUSSION THREAD

**Discussion by ICNSK** [SPECFEM/specfem3d#1715] [SPECFEM2D General]
Hi,
I am brand new user and trying to learn how to run xspecfem2D my research work. In the past I used finite difference simulators, which were limited to acoustic mode. I have input model for finite difference simulations, which is regularly samples binary file N1 x N2 and not sure how to convert it to mesh and model which could be used in xspecfem2D.

I was planning to use xmeshfem2D but could not find an example of how to use it with external mode.

Any information will be greatly appreciated, Thank you in advance

**Comment by homnath** [comment]
Hello,

Is your finite difference model homogeneous or heterogeneous? Specfem2D cannot directly take the finite difference model input. However, you can construct similar model. A few years ago, I wrote a simple program which can convert finite difference models into specfem2d or specfem3d files.
https://github.com/homnath/MeshAssist/blob/master/src/vti2cell.c
But depending on your model, it may be easier to construct a model using xmeshfem2D.
Please take a look at EXAMPLES/ folder.

Best,
Hom Nath


---------------------------------------
Hom Nath Gharti, PhD
Assistant Professor | Digital Earth Scientist
https://www.digitalearthscience.com/
Department of Geological Sciences and Geological Engineering
Miller Hall 314, 36 Union St
Queen’s University
Kingston, ON K7L 3N6, Canada


Queen’s University is situated on traditional
Anishinaabe and Haudenosaunee Territory.<https://www.queensu.ca/encyclopedia/t/traditional-territories>
________________________________
From: ICNSK ***@***.***>
Sent: Friday, April 19, 2024 8:15 PM
To: SPECFEM/specfem2d ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [SPECFEM/specfem2d] How to use xmeshfem2D with external model ? (Discussion SPECFEM/specfem2d#1219)


Hi,
I am brand new user and trying to learn how to run xspecfem2D my research work. In the past I used finite difference simulators, which were limited to acoustic mode. I have input model for finite difference simulations, which is regularly samples binary file N1 x N2 and not sure how to convert it to mesh and model which could be used in xspecfem2D.

I was planning to use xmeshfem2D but could not find an example of how to use it with external mode.

Any information will be greatly appreciated, Thank you in advance

**Reply by ICNSK** [reply]
Dear Hom Nath, 

Thank you for reply ! My finite difference model is heterogeneous.  In the EXAMPLES folder i found Marmousi mesh (from your previous work 2016/2023 ./EXAMPLES/Marmousi_mesh_of_the_model ) which could be a good starting point,   but i was not able to locate  PAR files with example how to use it.  I have no experiences with mesh generation software, and i was hoping to bypass that step at first  by using **specfem2d** functionality to create and read property files.  Below are the conceptual steps i am considering for my first try.  Please let me know if they makes any sense.

1.  Create placeholder model using PAR file with dummy material properties and simple topography  file. Model extent will correspond to extent of finite difference model ( key parameters are below). 
```
MODEL = default
SAVE_MODEL = ascii
nbmodels = 1
1 1 2700.d0 3000.d0 1732.051d0 0 0 9999 9999 0 0 0 0 0 0
interfacesfile = simple_topo.topo (will have 1 layer with nz=30 elements)
nx=20
nbregions = 1
1	20 1 30 1
```
2. Than use **xmeshfem2D** which will create 20x30 mesh  and use **specfem2d** to generate corresponding  **./DATA/proc000000_rho_vp_vs.dat**

3. Replace dummy values of rho, vp, vs by corresponding values from finite difference model based on XZ coordinates.
4. Run  **specfem2d** with **MODEL = ascii** to perform simulation using updated **./DATA/proc000000_rho_vp_vs.dat**

**Comment by homnath** [comment]
Can you try the example:
EXAMPLES/simple_topography_and_also_a_simple_fluid_layer
It uses an internal mesher.

Best,
Hom Nath


---------------------------------------
Hom Nath Gharti, PhD
Assistant Professor | Digital Earth Scientist
https://www.digitalearthscience.com/
Department of Geological Sciences and Geological Engineering
Miller Hall 314, 36 Union St
Queen’s University
Kingston, ON K7L 3N6, Canada


Queen’s University is situated on traditional
Anishinaabe and Haudenosaunee Territory.<https://www.queensu.ca/encyclopedia/t/traditional-territories>
________________________________

**Reply by ICNSK** [reply]
Dear @homnath,

Thank you for pointing me to a good example !  I was able to modify it and follow 2 step approach  to update material properties and run simulation with  [Marmousi Elastic Model](https://www.geoazur.fr/WIND/bin/view/Main/Data/WebHome) using rectangular grid produced by `xmeshfem2D`
![image](https://github.com/SPECFEM/specfem2d/assets/167325492/d73ff6e9-58e5-4a92-8fab-b852c4f41e9b)
![image](https://github.com/SPECFEM/specfem2d/assets/167325492/b0c8ca74-37c1-4566-8501-37f75ee6fe51)

**Comment by homnath** [comment]
Great! Congratulations!
________________________________
